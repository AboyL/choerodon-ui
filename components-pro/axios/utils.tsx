import buildURL from 'axios/lib/helpers/buildURL';

export function buildSortedURL(...args: any[]) {
  const builtURL = buildURL(...args);

  const [urlPath, queryString] = builtURL.split('?');

  if (queryString) {
    const paramsPair = queryString.split('&');
    return `${urlPath}?${paramsPair.sort().join('&')}`;
  }

  return builtURL;
}

export function buildURLWithData(data, ...args: any[]) {
  const builtURL = buildSortedURL(...args);
  if (data) {
    return `${builtURL}|${JSON.stringify(data)}`;
  }
  return builtURL;
}

export function isCacheLike(cache: any) {
  return !!(
    cache.set &&
    cache.get &&
    cache.del &&
    typeof cache.get === 'function' &&
    typeof cache.set === 'function' &&
    typeof cache.del === 'function'
  );
}
