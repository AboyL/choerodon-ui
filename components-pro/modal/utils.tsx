import { isValidElement, ReactNode } from 'react';
import { isArrayLike } from 'mobx';
import isNil from 'lodash/isNil';
import isBoolean from 'lodash/isBoolean';
import isNumber from 'lodash/isNumber';
import isString from 'lodash/isString';
import { ModalProps } from './Modal';

export type confirmProps = { iconType?: string; type?: string; children?: any };

export function normalizeProps(
  props: ModalProps & confirmProps | ReactNode,
): ModalProps & confirmProps {
  if (
    isBoolean(props) ||
    isNil(props) ||
    isNumber(props) ||
    isString(props) ||
    isValidElement(props) ||
    isArrayLike(props)
  ) {
    return {
      children: props,
    };
  }
  return props;
}
