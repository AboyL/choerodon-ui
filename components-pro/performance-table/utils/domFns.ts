// @ts-ignore
export function addEvent(el?: any, event: string, handler: EventListener | EventListenerObject | null, inputOptions?: Object): void {
  if (!el) return;
  const options = { capture: true, ...inputOptions };
  if (el.addEventListener) {
    el.addEventListener(event, handler, options);
  } else if (el.attachEvent) {
    el.attachEvent('on' + event, handler);
  } else {
    // $FlowIgnore: Doesn't think elements are indexable
    el['on' + event] = handler;
  }
}

export function removeEvent(el: any, event: string, handler: Function, inputOptions?: Object): void {
  if (!el) return;
  const options = { capture: true, ...inputOptions };
  if (el.removeEventListener) {
    el.removeEventListener(event, handler, options);
  } else if (el.detachEvent) {
    el.detachEvent('on' + event, handler);
  } else {
    // $FlowIgnore: Doesn't think elements are indexable
    el['on' + event] = null;
  }
}
