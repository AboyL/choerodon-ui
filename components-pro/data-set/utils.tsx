import queryString from 'querystringify';
import moment, { isDate, isMoment } from 'moment';
import { isArrayLike } from 'mobx';
import { AxiosRequestConfig } from 'axios';
import isBoolean from 'lodash/isBoolean';
import isObject from 'lodash/isObject';
import isString from 'lodash/isString';
import isArray from 'lodash/isArray';
import isNumber from 'lodash/isNumber';
import warning from 'choerodon-ui/lib/_util/warning';
import { getConfig } from 'choerodon-ui/lib/configure';
import Field, { FieldProps, Fields } from './Field';
import { BooleanValue, FieldType, RecordStatus, SortOrder } from './enum';
import DataSet from './DataSet';
import Record from './Record';
import Constants from './Constants';
import isEmpty from '../_util/isEmpty';
import * as ObjectChainValue from '../_util/ObjectChainValue';
import localeContext, { $l } from '../locale-context';
import { SubmitTypes, TransportType, TransportTypes } from './Transport';
import formatString from '../formatter/formatString';
import { Lang } from '../locale-context/enum';
import formatNumber from '../formatter/formatNumber';
import formatCurrency from '../formatter/formatCurrency';

export function append(url: string, suffix?: object) {
  if (suffix) {
    return url + queryString.stringify(suffix, url.indexOf('?') === -1);
  }
  return url;
}

export function getOrderFields(fields: Fields): Field[] {
  return [...fields.values()].filter(({ order }) => order);
}

export function processToJSON(value) {
  if (isDate(value)) {
    value = moment(value);
  }
  if (isMoment(value)) {
    value = value.format(Constants.DATE_JSON_FORMAT);
  }
  return value;
}

function processOne(value: any, field: Field, checkRange: boolean = true) {
  if (!isEmpty(value)) {
    const range = field.get('range');
    if (range && checkRange) {
      if (isArrayLike(range)) {
        if (isObject(value)) {
          const [start, end] = range;
          value[start] = processOne(value[start], field, false);
          value[end] = processOne(value[end], field, false);
        }
      } else if (isArrayLike(value)) {
        value[0] = processOne(value[0], field, false);
        value[1] = processOne(value[1], field, false);
      }
    } else if (value instanceof Date) {
      value = moment(value, Constants.DATE_JSON_FORMAT);
    } else if (!isObject(value)) {
      value = formatString(value, {
        trim: field.get('trim'),
        format: field.get('format'),
      });
      switch (field.type) {
        case FieldType.boolean: {
          const trueValue = field.get(BooleanValue.trueValue);
          const falseValue = field.get(BooleanValue.falseValue);
          if (value !== trueValue) {
            value = falseValue;
          }
          break;
        }
        case FieldType.number:
        case FieldType.currency:
          if (!isNaN(value)) {
            value = Number(value);
          } else {
            value = undefined;
          }
          break;
        case FieldType.string:
        case FieldType.intl:
        case FieldType.email:
        case FieldType.url:
          value = String(value);
          break;
        case FieldType.date:
        case FieldType.dateTime:
        case FieldType.time:
        case FieldType.week:
        case FieldType.month:
        case FieldType.year:
          value = moment(value, Constants.DATE_JSON_FORMAT);
          break;
        default:
      }
    }
  }
  return value;
}

export function processValue(value: any, field?: Field): any {
  if (field) {
    const multiple = field.get('multiple');
    const range = field.get('range');
    if (multiple) {
      if (isEmpty(value)) {
        value = [];
      } else if (!isArray(value)) {
        if (isString(multiple) && isString(value)) {
          value = value.split(multiple);
        } else {
          value = [value];
        }
      }
    }
    if (isArray(value) && (multiple || !range)) {
      return value.map(item => processOne(item, field));
    }
    return processOne(value, field);
  }
  return value;
}

export function childrenInfoForDelete(json: {}, children: { [key: string]: DataSet }): {} {
  return Object.keys(children).reduce((data, name) => {
    const child = children[name];
    if (child) {
      data[name] = [childrenInfoForDelete({}, child.children)];
    }
    return data;
  }, json);
}

export function sortTree(children: Record[], orderField: Field): Record[] {
  if (orderField && children.length > 0) {
    const { name, order } = orderField;
    const m = Number.MIN_SAFE_INTEGER;
    children.sort((record1, record2) => {
      const a = record1.get(name) || m;
      const b = record2.get(name) || m;
      if (isString(a) || isString(b)) {
        return order === SortOrder.asc
          ? String(a).localeCompare(String(b))
          : String(b).localeCompare(String(a));
      }
      return order === SortOrder.asc ? a - b : b - a;
    });
  }
  return children;
}

export function checkParentByInsert({ parent }: DataSet) {
  if (parent && !parent.current) {
    throw new Error($l('DataSet', 'cannot_add_record_when_head_no_current'));
  }
}

function getValueType(value: any): FieldType {
  return isBoolean(value)
    ? FieldType.boolean
    : isNumber(value)
    ? FieldType.number
    : isString(value)
    ? FieldType.string
    : isMoment(value)
    ? FieldType.date
    : isObject(value)
    ? FieldType.object
    : FieldType.auto;
}

function getBaseType(type: FieldType): FieldType {
  switch (type) {
    case FieldType.number:
    case FieldType.currency:
      return FieldType.number;
    case FieldType.dateTime:
    case FieldType.time:
    case FieldType.week:
    case FieldType.month:
    case FieldType.year:
      return FieldType.date;
    case FieldType.intl:
    case FieldType.url:
    case FieldType.email:
      return FieldType.string;
    default:
      return type;
  }
}

export function checkFieldType(value: any, field: Field): boolean {
  if (process.env.NODE_ENV !== 'production') {
    if (!isEmpty(value)) {
      if (isArrayLike(value)) {
        return value.every(item => checkFieldType(item, field));
      }
      const fieldType = getBaseType(field.type);
      const valueType =
        field.type === FieldType.boolean &&
        [field.get(BooleanValue.trueValue), field.get(BooleanValue.falseValue)].includes(value)
          ? FieldType.boolean
          : getValueType(value);
      if (
        fieldType !== FieldType.auto &&
        fieldType !== FieldType.reactNode &&
        fieldType !== valueType
      ) {
        warning(
          false,
          `Value type error: The value<${value}>'s type is ${valueType}, but the field<${field.name}>'s type is ${fieldType}.`,
        );
        return false;
      }
    }
  }
  return true;
}

let iframe;

export function doExport(url, data, method = 'post') {
  if (!iframe) {
    iframe = document.createElement('iframe');
    iframe.id = '_export_window';
    iframe.name = '_export_window';
    iframe.style.cssText =
      'position:absolute;left:-10000px;top:-10000px;width:1px;height:1px;display:none';
    document.body.appendChild(iframe);
  }

  const form = document.createElement('form');
  form.target = '_export_window';
  form.method = method;
  form.action = url;
  const s = document.createElement('input');
  s.id = '_request_data';
  s.type = 'hidden';
  s.name = '_request_data';
  s.value = JSON.stringify(data);
  form.appendChild(s);
  document.body.appendChild(form);
  form.submit();
  document.body.removeChild(form);
}

export function findBindFields(myField: Field, fields: Fields, excludeSelf?: boolean): Field[] {
  const { name } = myField;
  return [...fields.values()].filter(field => {
    if (field !== myField) {
      const bind = field.get('bind');
      return isString(bind) && bind.startsWith(`${name}.`);
    }
    return !excludeSelf;
  });
}

export function findBindField(
  myField: Field | string,
  fields: Fields,
  callback?: (field: Field) => boolean,
): Field | undefined {
  const name = isString(myField) ? myField : myField.name;
  return [...fields.values()].find(field => {
    if (field.name !== name) {
      const bind = field.get('bind');
      return isString(bind) && bind.startsWith(`${name}.`) && (!callback || callback(field));
    }
    return false;
  });
}

function numberSorter(a, b) {
  return a - b;
}

function stringSorter(a, b) {
  return String(a || '').localeCompare(String(b || ''));
}

export function getFieldSorter(field: Field) {
  const { name } = field;

  switch (field.type) {
    case FieldType.number:
    case FieldType.currency:
    case FieldType.date:
    case FieldType.dateTime:
    case FieldType.week:
    case FieldType.month:
    case FieldType.year:
    case FieldType.time:
      return field.order === SortOrder.asc
        ? (a, b) => numberSorter(a.get(name), b.get(name))
        : (a, b) => numberSorter(b.get(name), a.get(name));
    default:
      return field.order === SortOrder.asc
        ? (a, b) => stringSorter(a.get(name), b.get(name))
        : (a, b) => stringSorter(b.get(name), a.get(name));
  }
}

export function generateJSONData(
  array: object[],
  record: Record,
  isSelect?: boolean,
  noCascade?: boolean,
) {
  const json = record.toJSONData(noCascade);
  if (json.__dirty || isSelect) {
    delete json.__dirty;
    array.push(json);
  }
}

export function prepareSubmitData(
  records: Record[],
  isSelect?: boolean,
  noCascade?: boolean,
): [object[], object[], object[], object[]] {
  const created: object[] = [];
  const updated: object[] = [];
  const destroyed: object[] = [];
  const cascade: object[] = [];

  function storeWith(status) {
    switch (status) {
      case RecordStatus.add:
        return created;
      case RecordStatus.update:
        return updated;
      case RecordStatus.delete:
        return destroyed;
      default:
        return cascade;
    }
  }

  records.forEach(
    record =>
      (noCascade && record.status === RecordStatus.sync) ||
      generateJSONData(storeWith(record.status), record, isSelect, noCascade),
  );
  return [created, updated, destroyed, cascade];
}

function defaultAxiosConfigAdapter(config: AxiosRequestConfig): AxiosRequestConfig {
  return config;
}

function generateConfig(
  config: TransportType,
  dataSet: DataSet,
  data?: any,
  params?: any,
): AxiosRequestConfig {
  if (isString(config)) {
    return {
      url: config,
    };
  }
  if (typeof config === 'function') {
    return config({ data, dataSet, params });
  }
  return config;
}

export function axiosConfigAdapter(
  type: TransportTypes,
  dataSet: DataSet,
  data?: any,
  params?: any,
): AxiosRequestConfig {
  const newConfig: AxiosRequestConfig = {
    data,
    params,
    method: 'post',
  };
  const { [type]: globalConfig, adapter: globalAdapter = defaultAxiosConfigAdapter } =
    getConfig('transport') || {};
  const { [type]: config, adapter } = dataSet.transport;
  if (globalConfig) {
    Object.assign(newConfig, generateConfig(globalConfig, dataSet, data, params));
  }
  if (config) {
    Object.assign(newConfig, generateConfig(config, dataSet, data, params));
  }
  if (newConfig.data && newConfig.method && newConfig.method.toLowerCase() === 'get') {
    newConfig.params = {
      ...newConfig.params,
      ...newConfig.data,
    };
  }
  return (adapter || globalAdapter)(newConfig, type) || newConfig;
}

export function prepareForSubmit(
  type: SubmitTypes,
  data: object[],
  configs: AxiosRequestConfig[],
  dataSet: DataSet,
): object[] {
  if (data.length) {
    const newConfig = axiosConfigAdapter(type, dataSet, data);
    if (newConfig.url) {
      configs.push(newConfig);
    } else {
      return data;
    }
  }
  return [];
}

export function generateResponseData(item: any, dataKey?: string): object[] {
  if (item) {
    if (isArray(item)) {
      return item;
    }
    if (isObject(item)) {
      if (dataKey) {
        const result = ObjectChainValue.get(item, dataKey);
        if (result === undefined) {
          return [item];
        }
        if (isArray(result)) {
          return result;
        }
        if (isObject(result)) {
          return [result];
        }
      } else {
        return [item];
      }
    }
  }
  return [];
}

export function getRecordValue(
  data: any,
  cb: (record: Record, fieldName: string) => boolean,
  fieldName?: string,
) {
  if (fieldName) {
    const field = this.getField(fieldName);
    if (field) {
      const bind = field.get('bind');
      if (bind) {
        fieldName = bind;
      }
    }
    const { dataSet } = this;
    if (dataSet) {
      const { checkField } = dataSet.props;
      if (checkField && checkField === fieldName) {
        const trueValue = field ? field.get(BooleanValue.trueValue) : true;
        const falseValue = field ? field.get(BooleanValue.falseValue) : false;
        const { children } = this;
        if (children) {
          return children.every(child => cb(child, checkField) === trueValue)
            ? trueValue
            : falseValue;
        }
      }
    }
    return ObjectChainValue.get(data, fieldName as string);
  }
}

export function processIntlField(
  name: string,
  fieldProps: FieldProps,
  callback: (name: string, props: FieldProps) => Field,
  dataSet?: DataSet,
): Field {
  const tlsKey = getConfig('tlsKey');
  const { supports } = localeContext;
  const languages = Object.keys(supports);
  const { type, dynamicProps } = fieldProps;
  if (type === FieldType.intl) {
    languages.forEach(language =>
      callback(`${tlsKey}.${name}.${language}`, {
        type: FieldType.string,
        label: `${supports[language]}`,
      }),
    );
    const { lang = localeContext.locale.lang } = dataSet || {};
    const newDynamicProps =
      typeof dynamicProps === 'function'
        ? props => {
            return {
              ...dynamicProps(props),
              bind: props.record.get(tlsKey) ? `${tlsKey}.${name}.${lang}` : undefined,
            };
          }
        : {
            ...dynamicProps,
            bind: props => {
              return props.record.get(tlsKey) ? `${tlsKey}.${name}.${lang}` : undefined;
            },
          };
    return callback(name, {
      ...fieldProps,
      dynamicProps: newDynamicProps,
    });
  }
  return callback(name, fieldProps);
}

export function findBindFieldBy(myField: Field, fields: Fields, prop: string): Field | undefined {
  const value = myField.get(prop);
  const myName = myField.name;
  return [...fields.values()].find(field => {
    const bind = field.get('bind');
    return bind && bind === `${myName}.${value}`;
  });
}

export function processFieldValue(value, field: Field, lang: Lang, showValueIfNotFound?: boolean) {
  const { type } = field;
  if (type === FieldType.number) {
    return formatNumber(value, lang);
  }
  if (type === FieldType.currency) {
    return formatCurrency(value, lang, {
      currency: field.get('currency'),
    });
  }
  return field.getText(value, showValueIfNotFound);
}
