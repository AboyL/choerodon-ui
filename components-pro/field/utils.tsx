import isObject from 'lodash/isObject';
import isNil from 'lodash/isNil';
import { isArrayLike } from 'mobx';
import { FieldType } from '../data-set/enum';
import Constants from '../data-set/Constants';
import Field from '../data-set/Field';

export function toRangeValue(value: any, range?: boolean | [string, string]): [any, any] {
  if (isArrayLike(range)) {
    if (isObject(value)) {
      const [start, end] = range;
      return [value[start], value[end]];
    }
  } else if (isArrayLike(value)) {
    return value.slice(0, 2) as [any, any];
  }
  return [undefined, undefined];
}

export function fromRangeValue(value: any[], range?: boolean | [string, string]): any {
  if (isArrayLike(range)) {
    const [start, end] = range;
    return {
      [start]: value[0],
      [end]: value[1],
    };
  }
  return value;
}

export function toMultipleValue(value: any, range?: boolean | [string, string]) {
  if (!isNil(value)) {
    const multipleValue = isArrayLike(value) ? value.slice() : [value];
    if (range) {
      return multipleValue.map(item => toRangeValue(item, range));
    }
    return multipleValue;
  }
  return [];
}

export function getDateFormatByFieldType(type: FieldType) {
  switch (type) {
    case FieldType.date:
      return Constants.DATE_FORMAT;
    case FieldType.dateTime:
      return Constants.DATE_TIME_FORMAT;
    case FieldType.week:
      return Constants.WEEK_FORMAT;
    case FieldType.month:
      return Constants.MONTH_FORMAT;
    case FieldType.year:
      return Constants.YEAR_FORMAT;
    case FieldType.time:
      return Constants.TIME_FORMAT;
    default:
      return Constants.DATE_FORMAT;
  }
}

export function getDateFormatByField(field?: Field, type?: FieldType): string {
  if (field) {
    return field.get('format') || getDateFormatByFieldType(type || field.type);
  }
  if (type) {
    return getDateFormatByFieldType(type);
  }
  return Constants.DATE_JSON_FORMAT;
}
