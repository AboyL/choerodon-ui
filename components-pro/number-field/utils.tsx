import { FieldType } from '../data-set/enum';
import normalizeLanguage from '../_util/normalizeLanguage';

export const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 2 ** 53 - 1;

export function getPrecision(value: number): number {
  const valueString = value.toString();
  if (valueString.indexOf('e-') >= 0) {
    return parseInt(valueString.slice(valueString.indexOf('e-') + 2), 10);
  }
  if (valueString.indexOf('.') >= 0) {
    return valueString.length - valueString.indexOf('.') - 1;
  }
  return 0;
}

function getMaxPrecision(value: number, step: number): number {
  const stepPrecision = getPrecision(step);
  const currentValuePrecision = getPrecision(value);
  if (!value) {
    return stepPrecision;
  }
  return Math.max(currentValuePrecision, stepPrecision);
}

function getPrecisionFactor(value: number, step: number): number {
  return 10 ** getMaxPrecision(value, step);
}

function precisionFix(value: number, precisionFactor: number): number {
  return Math.round(value * precisionFactor);
}

export function plus(...values: number[]) {
  if (values.length > 2) {
    return plus(values.shift() as number, plus(...values));
  }
  if (values.length < 2) {
    return values[0];
  }
  const v1 = values[0];
  const v2 = values[1];
  const precisionFactor = getPrecisionFactor(v1, v2);
  return (precisionFix(v1, precisionFactor) + precisionFix(v2, precisionFactor)) / precisionFactor;
}

function getBeforeStepValue(value: number, minFactor: number, stepFactor: number): number {
  return value - ((value - minFactor) % stepFactor);
}

export function getNearStepValues(
  value: number,
  step: number,
  min: number = -MAX_SAFE_INTEGER,
  max: number = MAX_SAFE_INTEGER,
): number[] | undefined {
  const precisionFactor = getPrecisionFactor(value, step);
  const valueFactor = precisionFix(value, precisionFactor);
  const minFactor = precisionFix(min, precisionFactor);
  const minFactorBase = min === -MAX_SAFE_INTEGER ? 0 : minFactor;
  const maxFactor = precisionFix(max, precisionFactor);
  const stepFactor = precisionFix(step, precisionFactor);
  let beforeStepFactor = getBeforeStepValue(valueFactor, minFactorBase, stepFactor);
  if (beforeStepFactor === valueFactor) {
    return undefined;
  }
  if (beforeStepFactor > maxFactor) {
    beforeStepFactor = getBeforeStepValue(maxFactor, minFactorBase, stepFactor);
  } else if (beforeStepFactor < minFactor) {
    beforeStepFactor = minFactor;
  }
  const afterStepFactor = beforeStepFactor + stepFactor;
  const values: number[] = [beforeStepFactor / precisionFactor];
  if (afterStepFactor <= maxFactor) {
    values.push(afterStepFactor / precisionFactor);
  }
  return values;
}

let supportsLocales;

export function toLocaleStringSupportsLocales() {
  if (supportsLocales === undefined) {
    try {
      (0).toLocaleString('i');
      supportsLocales = false;
    } catch (e) {
      supportsLocales = e.name === 'RangeError';
    }
  }
  return supportsLocales;
}

function getNumberFormatOptions(
  type: FieldType,
  options?: Intl.NumberFormatOptions,
): Intl.NumberFormatOptions {
  if (type === FieldType.number) {
    return { style: 'decimal' };
  }
  if (options && options.currency) {
    return { style: 'currency' };
  }
  return { style: 'decimal', minimumFractionDigits: 2, maximumFractionDigits: 2 };
}

function toLocaleStringPolyfill(
  value: number,
  type: FieldType,
  options?: Intl.NumberFormatOptions,
) {
  if (type === FieldType.number) {
    const fraction = String(value).split('.')[1];
    return value.toLocaleString().split('.')[0] + (fraction ? `.${fraction}` : '');
  }
  const currency = options && options.currency;
  return `${currency ? `${currency} ` : ''}${value.toLocaleString()}`;
}

export function formatNumber(value, lang: string, options?: Intl.NumberFormatOptions) {
  const v = parseFloat(value);
  if (!isNaN(v)) {
    if (toLocaleStringSupportsLocales()) {
      return v.toLocaleString(normalizeLanguage(lang), {
        ...getNumberFormatOptions(FieldType.number, options),
        ...options,
      });
    }
    return toLocaleStringPolyfill(v, FieldType.number, options);
  }
  return value;
}

export function formatCurrency(value, lang: string, options?: Intl.NumberFormatOptions) {
  const v = parseFloat(value);
  if (!isNaN(v)) {
    if (toLocaleStringSupportsLocales()) {
      return v.toLocaleString(normalizeLanguage(lang), {
        ...getNumberFormatOptions(FieldType.currency, options),
        ...options,
      });
    }
    return toLocaleStringPolyfill(v, FieldType.currency, options);
  }
  return value;
}
