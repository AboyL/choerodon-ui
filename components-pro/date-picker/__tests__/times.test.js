import React from 'react';
import { mount } from 'enzyme';
import moment from 'moment';
import TimesPicker from '../../time-picker';
import focusTest from '../../../tests/shared/focusTest';
import { disableWrapper } from './utils';

describe('times-picker-pro', () => {
  focusTest(TimesPicker);

  it('the times will be controlled by the value', () => {
    const wrapper = mount(<TimesPicker value={moment('2020-02-20 20:20:19')} />);
    expect(
      wrapper
        .find('input')
        .at(0)
        .prop('value'),
    ).toBe('20:20:19');
    wrapper.setProps({ value: moment('2020-02-21 20:22:18') });
    wrapper.update();
    expect(
      wrapper
        .find('input')
        .at(0)
        .prop('value'),
    ).toBe('20:22:18');
  });

  it('should has disabled property can not do anything', () => {
    const wrapper = mount(<TimesPicker />);
    disableWrapper(wrapper);
  });
});
