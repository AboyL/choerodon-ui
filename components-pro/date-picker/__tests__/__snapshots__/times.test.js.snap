// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`times-picker-pro renders timePicker correctly 1`] = `
<TimePicker
  autoComplete="off"
  clearButton={true}
  mode="time"
  multiple={false}
  noValidate={false}
  readOnly={false}
  showHelp="newLine"
  suffixCls="calendar-picker"
  trigger={
    Array [
      "focus",
      "click",
    ]
  }
  triggerHiddenDelay={50}
  triggerShowDelay={150}
  trim="both"
>
  <Tooltip
    arrowPointAtCenter={false}
    autoAdjustOverflow={true}
    defaultHidden={true}
    mouseEnterDelay={100}
    mouseLeaveDelay={100}
    placement="bottomLeft"
    suffixCls="tooltip"
    theme="light"
    title={null}
    transitionName="zoom-big-fast"
    trigger={
      Array [
        "hover",
      ]
    }
  >
    <Trigger
      action={
        Array [
          "hover",
        ]
      }
      arrowPointAtCenter={false}
      autoAdjustOverflow={true}
      blurDelay={0}
      builtinPlacements={
        Object {
          "bottom": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tc",
              "bc",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "bottomLeft": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "bottomRight": Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tr",
              "br",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "left": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "cr",
              "cl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "leftBottom": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "br",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "leftTop": Object {
            "offset": Array [
              -10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tr",
              "tl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "right": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "cl",
              "cr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "rightBottom": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bl",
              "br",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "rightTop": Object {
            "offset": Array [
              10,
              0,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "tr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "top": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bc",
              "tc",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "topLeft": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "bl",
              "tl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
          "topRight": Object {
            "offset": Array [
              0,
              -10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "br",
              "tr",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          },
        }
      }
      defaultPopupHidden={true}
      focusDelay={150}
      mouseEnterDelay={100}
      mouseLeaveDelay={100}
      popupContent={null}
      popupPlacement="bottomLeft"
      prefixCls="c7n-pro-tooltip"
      suffixCls="tooltip"
      theme="light"
      title={null}
      transitionName="zoom-big-fast"
    >
      <TriggerChild
        key=".$.$element"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <span
          className="c7n-pro-calendar-picker-wrapper c7n-pro-calendar-picker-empty c7n-pro-calendar-picker-suffix-button"
          key=".$element"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <label
            onMouseDown={[Function]}
          >
            <Trigger
              action={
                Array [
                  "focus",
                  "click",
                ]
              }
              blurDelay={50}
              builtinPlacements={
                Object {
                  "bottomLeft": Object {
                    "offset": Array [
                      0,
                      4,
                    ],
                    "overflow": Object {
                      "adjustX": 1,
                      "adjustY": 1,
                    },
                    "points": Array [
                      "tl",
                      "bl",
                    ],
                  },
                  "bottomRight": Object {
                    "offset": Array [
                      0,
                      4,
                    ],
                    "overflow": Object {
                      "adjustX": 1,
                      "adjustY": 1,
                    },
                    "points": Array [
                      "tr",
                      "br",
                    ],
                  },
                  "topLeft": Object {
                    "offset": Array [
                      0,
                      -4,
                    ],
                    "overflow": Object {
                      "adjustX": 1,
                      "adjustY": 1,
                    },
                    "points": Array [
                      "bl",
                      "tl",
                    ],
                  },
                  "topRight": Object {
                    "offset": Array [
                      0,
                      -4,
                    ],
                    "overflow": Object {
                      "adjustX": 1,
                      "adjustY": 1,
                    },
                    "points": Array [
                      "br",
                      "tr",
                    ],
                  },
                }
              }
              defaultPopupHidden={true}
              focusDelay={150}
              getPopupStyleFromAlign={[Function]}
              getRootDomNode={[Function]}
              mouseEnterDelay={150}
              mouseLeaveDelay={50}
              onPopupAnimateAppear={[Function]}
              onPopupAnimateEnd={[Function]}
              onPopupHiddenChange={[Function]}
              popupContent={
                <TimesView
                  date={"2020-02-19T16:00:00.000Z"}
                  isValidDate={[Function]}
                  mode="time"
                  onSelect={[Function]}
                  onSelectedDateChange={[Function]}
                  onViewModeChange={[Function]}
                  suffixCls="calendar"
                />
              }
              popupHidden={true}
              popupPlacement="bottomLeft"
              prefixCls="c7n-pro-calendar-picker"
              transitionName="slide-up"
            >
              <TriggerChild
                key=".$text"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
              >
                <input
                  autoComplete="off"
                  className="c7n-pro-calendar-picker"
                  key="text"
                  multiple={false}
                  noValidate={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  readOnly={false}
                  type="text"
                  value=""
                />
              </TriggerChild>
              <Popup
                align={
                  Object {
                    "offset": Array [
                      0,
                      4,
                    ],
                    "overflow": Object {
                      "adjustX": 1,
                      "adjustY": 1,
                    },
                    "points": Array [
                      "tl",
                      "bl",
                    ],
                  }
                }
                className="c7n-pro-calendar-picker-popup"
                getClassNameFromAlign={[Function]}
                getRootDomNode={[Function]}
                getStyleFromAlign={[Function]}
                hidden={true}
                key="popup"
                onAnimateAppear={[Function]}
                onAnimateEnd={[Function]}
                onMouseDown={[Function]}
                suffixCls="popup"
                transitionName="slide-up"
              />
            </Trigger>
            <div
              className="c7n-pro-calendar-picker-inner-button c7n-pro-calendar-picker-clear-button"
              key="inner-button"
            >
              <Icon
                onClick={[Function]}
                type="close"
              >
                <i
                  className="icon icon-close"
                  onClick={[Function]}
                />
              </Icon>
            </div>
            <div
              className="c7n-pro-calendar-picker-suffix"
              onMouseDown={[Function]}
            >
              <Icon
                className="c7n-pro-calendar-picker-trigger"
                type="date_range"
              >
                <i
                  className="icon icon-date_range c7n-pro-calendar-picker-trigger"
                />
              </Icon>
            </div>
          </label>
        </span>
      </TriggerChild>
      <Popup
        align={
          Object {
            "offset": Array [
              0,
              10,
            ],
            "overflow": Object {
              "adjustX": 1,
              "adjustY": 1,
            },
            "points": Array [
              "tl",
              "bl",
            ],
            "targetOffset": Array [
              0,
              0,
            ],
          }
        }
        className="c7n-pro-tooltip-popup"
        getClassNameFromAlign={[Function]}
        getRootDomNode={[Function]}
        hidden={true}
        key="popup"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        suffixCls="popup"
        transitionName="zoom-big-fast"
      />
    </Trigger>
  </Tooltip>
</TimePicker>
`;
